src/Nethermind/Nethermind.Optimism/CL/L1Bridge/EthereumL1Bridge.cs

```diff
@@ -0,0 +1,293 @@
+    {
+        const ulong l1SlotTime = 12;
+        return (timestamp - _engineParameters.L1BeaconGenesisSlotTime!.Value) / l1SlotTime;
+    }
+
+    private async Task ProcessBlobBatcherTransaction(L1Transaction transaction, int startingBlobIndex, ulong slotNumber, CancellationToken cancellationToken)
+    {
+        BlobSidecar[] blobSidecars = await _beaconApi.GetBlobSidecars(slotNumber, startingBlobIndex,
+            startingBlobIndex + transaction.BlobVersionedHashes!.Length - 1, cancellationToken);
+
+        if (blobSidecars is null)
+        {
+            if (_logger.IsWarn)
+            {
+                string blobVersionedHashes = string.Join(',', transaction.BlobVersionedHashes.Select(hash => hash.ToHexString()));
+                _logger.Warn($"Failed to get blob sidecars for slot {slotNumber}. Transaction: {transaction.Hash}. BlobVersionedHashes: {blobVersionedHashes}");
+            }
+
+            throw new ArgumentNullException(nameof(blobSidecars));
+        }
+
+        for (int i = 0; i < transaction.BlobVersionedHashes.Length; i++)
+        {
+            await _decodingPipeline.DaDataWriter.WriteAsync(blobSidecars[i].Blob, cancellationToken);
+        }
+    }
+
+    private void ProcessCalldataBatcherTransaction(L1Transaction transaction)
+    {
+        if (_logger.IsError)
+        {
+            _logger.Error($"GOT REGULAR TRANSACTION");
+        }
+
+        throw new NotImplementedException();
+    }
+
+    private async Task TryUpdateFinalized()
+    {
+        if (_currentHeadNumber - _currentFinalizedNumber >= 64)
+        {
+            L1Block newFinalized = await GetFinalized();
+            SetFinalized(newFinalized);
+        }
+    }
+
+    private void SetFinalized(L1Block newFinalized)
+    {
+        if (newFinalized.Hash != _currentFinalizedHash)
+        {
+            _logger.Error($"New finalized head signal. Number: {newFinalized.Number}, Hash: {newFinalized.Hash}");
+            _currentFinalizedHash = newFinalized.Hash;
+            _currentFinalizedNumber = newFinalized.Number;
+        }
+    }
+
+    // Gets all blocks from range [{segmentStartNumber}, {headNumber})
+    private async Task RollBack(Hash256 headHash, ulong headNumber, ulong segmentStartNumber, CancellationToken cancellationToken)
+    {
+        Hash256 currentHash = headHash;
+        L1Block[] chainSegment = new L1Block[headNumber - segmentStartNumber + 1];
+        for (int i = chainSegment.Length - 1; i >= 0; i--)
+        {
+            _logger.Info($"Rolling back L1 head. {i} to go. Block {currentHash}");
+            chainSegment[i] = await GetBlockByHash(currentHash);
+            currentHash = chainSegment[i].ParentHash;
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
+        }
+
+        for (int i = 0; i < chainSegment.Length; i++)
+        {
+            await ProcessBlock(chainSegment[i], cancellationToken);
+        }
```

Comment:
I think it's safe to run all of these tasks in parallel:
```suggestion
        await Task.WhenAll(chainSegment.Select(s => ProcessBlock(s, cancellationToken)));
```

Reply:
No, we have to push blobs in order they appear